File: problems/0010_regular-expression-matching/backtrace/Solution.java
class Solution {
    public boolean isMatch(String s, String p) {
        final int sLen = s.length(), pLen = p.length();
        if (pLen == 0)
            return sLen == 0;
        final char p0 = p.charAt(0);
        final boolean matchP0 = (sLen > 0 &&
                (p0 == s.charAt(0) || p0 == '.'));
        if (pLen >= 2 && p.charAt(1) == '*')
            return (isMatch(s, p.substring(2)) ||
                    (matchP0 && isMatch(s.substring(1), p)));
        else
            return (matchP0 && isMatch(s.substring(1), p.substring(1)));
    }
}

File: problems/0010_regular-expression-matching/backtrace/solution.go
func isMatch(s string, p string) bool {
    sLen, pLen := len(s), len(p)
    if pLen == 0 {
        return (sLen == 0)
    }
    p0 := p[0]
    matchP0 := (sLen > 0 &&
            (p0 == s[0] || p0 == '.'))
    if pLen >= 2 && p[1] == '*' {
        return (isMatch(s, p[2:]) ||
                (matchP0 && isMatch(s[1:], p)))
    } else {
        return (matchP0 && isMatch(s[1:], p[1:]))
    }
}
